tests/test_interface.py

```diff
@@ -0,0 +1,138 @@
+PP_RETURN ['charge_history'] = 0.0
+PP_RETURN ['payment_id'] = 1
+PP_RETURN ['details']['is_linked'] = True
+
+CC_RETURN = dict(CREDIT)
+CC_RETURN['is_default'] = False
+CC_RETURN['charge_history'] = 0.0
+CC_RETURN['payment_id'] = 1
+
+#this should always be added second
+DC_RETURN = dict(DEBIT)
+DC_RETURN['is_default'] = False
+DC_RETURN['charge_history'] = 0.0
+DC_RETURN['payment_id'] = 2
+
+ps = PaymentService()
+
+class TestInterface(unittest.TestCase):
+
+    def setUp(self):
+        payments.app.debug = True
+        payments.app.config['SQLALCHEMY_DATABASE_URI'] = 'postgresql://payments:payments@localhost:5432/test'
+        app_db.create_all()  # make our sqlalchemy tables
+        
+        data = CREDIT
+
+        payment = Payment()
+        payment.deserialize(data)
+        app_db.session.add(payment)
+        app_db.session.commit()
+        self.app = payments.app.test_client()
+
+    def tearDown(self):
+        app_db.session.remove()
+        app_db.drop_all()
+
+    def test_interface_add_card_returns_json(self):
+        data = DEBIT
+        payment = ps.add_payment(data)
+        self.assertTrue(type(payment), type({}))
+        self.assertTrue(payment['nickname'] == 'my debit')
+        self.assertTrue(payment['details']['user_name'] == 'John Jameson')
+
+    def test_interface_add_card_to_db(self):
+        data = DEBIT
+        p2 = app_db.session.query(Payment).get(2)
+        self.assertEqual(p2, None)
+        ps.add_payment(data)
+        p2 = app_db.session.query(Payment).get(2)
+        d2 = p2.details
+        self.assertNotEqual(p2, None)
+        self.assertEqual(p2.nickname, 'my debit')
+        self.assertEqual(d2.user_name, 'John Jameson')
+
+    def test_interface_add_paypal_to_db(self):
+        data = PAYPAL
+        p2 = app_db.session.query(Payment).get(2)
+        self.assertEqual(p2, None)
+        ps.add_payment(data)
+        p2 = app_db.session.query(Payment).get(2)
+        d2 = p2.details
+        self.assertNotEqual(p2, None)
+        self.assertEqual(p2.nickname, 'my paypal')
+        self.assertEqual(d2.user_name, 'John Jameson')
+        self.assertEqual(d2.is_linked, True)
+
+    def test_interface_add_missing_details(self):
+        data = {'nickname' : 'my debit', 'user_id' : 2, 'payment_type' : 'debit'}
+        self.assertRaises(DataValidationError, ps.add_payment, data)
+
+    def test_interface_add_bad_data(self):
+        data = BAD_DATA
+        self.assertRaises(DataValidationError, ps.add_payment, data)
+
+    def test_interface_add_garbage(self):
+        garbage = 'afv@#(&@(#Z@#>X@C8rq rq34tr0q934r 9qr@(#*(@!$))'
+        self.assertRaises(DataValidationError, ps.add_payment, garbage)
+
+    """ 
+    adding some mock tests in addition to the above since add_payment
+    relies on functions in Payment / Detail
+
+    these also rely on Deserialize methods, but since those reuturn model objects,
+    I'm not sure how to mock those quite yet... will revisit
```

Comment:
Don't worry about what the return type of `deserialize` - that method's functionality should be mocked anyway since we just care about what `add_payment` is doing.  Fortunately the functionality is really straightforward.

What I recommend is that you add some assertions about how the mocked methods get called.  For instance, since you are passing in `CREDIT` for the `payment_data` parameter, you will want to make sure that the `serialize` method is called with this parameter:

```
mock_serial.assert_called_once_with(CREDIT)
```

We also want to make sure that no actual db interactions occur.  Fortunately the mocking is quite simple:

```
@patch('app.db.app_db')
```

This patches the module itself, so whenever `app_db` gets used, it'll just return MagicMocks.  You can make similar assertions on the `session` calls as the one I gave above.

Lastly, do another `assert_called_once_with` for a mocked `serialize` method.  Make sure that whatever the mocked `serialize` method returns is what the `add_payment` method returns.  As simple as it might be, that's what the unit test should assert.